File: app/src/main/java/org/ametro/ui/navigation/adapter/NavigationDrawerAdapter.java
package org.ametro.ui.navigation.adapter;

import android.content.Context;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.BaseAdapter;

import org.ametro.ui.navigation.entities.INavigationItemGroup;
import org.ametro.ui.navigation.entities.NavigationCheckBoxItem;
import org.ametro.ui.navigation.entities.NavigationHeader;
import org.ametro.ui.navigation.entities.NavigationItem;
import org.ametro.ui.navigation.entities.NavigationSplitter;
import org.ametro.ui.navigation.entities.NavigationSubHeader;
import org.ametro.ui.navigation.entities.NavigationTextItem;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class NavigationDrawerAdapter extends BaseAdapter {

    private final List<NavigationItem> items;
    private final LayoutInflater inflater;

    private final Map<Class, Integer> viewItemTypes = new HashMap<>();
    private final Map<Class, IHolderFactory> viewItemHolderFactories = new HashMap<>();


    public NavigationDrawerAdapter(Context context, NavigationItem[] items) {

        viewItemTypes.put(NavigationHeader.class, 0);
        viewItemTypes.put(NavigationTextItem.class, 1);
        viewItemTypes.put(NavigationSplitter.class, 2);
        viewItemTypes.put(NavigationSubHeader.class, 3);
        viewItemTypes.put(NavigationCheckBoxItem.class, 4);

        viewItemHolderFactories.put(NavigationHeader.class, new NavigationHeaderHolderFactory());
        viewItemHolderFactories.put(NavigationTextItem.class, new NavigationTextItemHolderFactory());
        viewItemHolderFactories.put(NavigationSplitter.class, new NavigationSplitterHolderFactory());
        viewItemHolderFactories.put(NavigationSubHeader.class, new NavigationSubHeaderHolderFactory());
        viewItemHolderFactories.put(NavigationCheckBoxItem.class, new NavigationCheckBoxItemHolderFactory());

        this.inflater = (LayoutInflater) context.getSystemService(Context.LAYOUT_INFLATER_SERVICE);
        this.items = flattenItems(items);
    }

    public void setNavigationItems(NavigationItem[] items) {
        this.items.clear();
        this.items.addAll(flattenItems(items));
        this.notifyDataSetChanged();
    }

    @Override
    public boolean areAllItemsEnabled() {
        return false;
    }

    @Override
    public boolean isEnabled(int position) {
        return items.get(position).isEnabled();
    }

    @Override
    public int getCount() {
        return items.size();
    }

    @Override
    public NavigationItem getItem(int position) {
        return items.get(position);
    }

    @Override
    public long getItemId(int position) {
        return position;
    }

    @Override
    public boolean hasStableIds() {
        return false;
    }

    @Override
    public View getView(int position, View convertView, ViewGroup parent) {
        IHolder holder;
        if (convertView == null) {
            IHolderFactory factory = viewItemHolderFactories.get(items.get(position).getClass());
            convertView = factory.createView(inflater, parent);
            holder = factory.createHolder(convertView);
        } else {
            holder = (IHolder) convertView.getTag();
        }
        holder.update(items.get(position));
        return convertView;
    }

    @Override
    public int getItemViewType(int position) {
        return viewItemTypes.get(items.get(position).getClass());
    }

    @Override
    public int getViewTypeCount() {
        return viewItemTypes.size();
    }

    @Override
    public boolean isEmpty() {
        return items.size() == 0;
    }

    private List<NavigationItem> flattenItems(NavigationItem[] items) {
        List<NavigationItem> flattenList = new ArrayList<>();
        for (NavigationItem item : items) {
            flattenList.add(item);
            if (item instanceof INavigationItemGroup) {
                INavigationItemGroup group = (INavigationItemGroup) item;
                flattenList.addAll(flattenItems(group.getItems()));
            }
        }
        return flattenList;
    }
}

File: app/src/main/java/io/github/romangolovanov/apps/ametro/ui/navigation/adapter/NavigationDrawerAdapter.kt
package io.github.romangolovanov.apps.ametro.ui.navigation.adapter

import android.content.Context
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.BaseAdapter
import io.github.romangolovanov.apps.ametro.ui.navigation.entities.INavigationItemGroup
import io.github.romangolovanov.apps.ametro.ui.navigation.entities.NavigationCheckBoxItem
import io.github.romangolovanov.apps.ametro.ui.navigation.entities.NavigationHeader
import io.github.romangolovanov.apps.ametro.ui.navigation.entities.NavigationItem
import io.github.romangolovanov.apps.ametro.ui.navigation.entities.NavigationSplitter
import io.github.romangolovanov.apps.ametro.ui.navigation.entities.NavigationSubHeader
import io.github.romangolovanov.apps.ametro.ui.navigation.entities.NavigationTextItem

class NavigationDrawerAdapter(context: Context, items: Array<NavigationItem>) : BaseAdapter() {

    private val items: MutableList<NavigationItem>
    private val inflater: LayoutInflater

    private val viewItemTypes: Map<Class<*>, Int> = mapOf(
        NavigationHeader::class.java to 0,
        NavigationTextItem::class.java to 1,
        NavigationSplitter::class.java to 2,
        NavigationSubHeader::class.java to 3,
        NavigationCheckBoxItem::class.java to 4
    )

    private val viewItemHolderFactories: Map<Class<*>, IHolderFactory> = mapOf(
        NavigationHeader::class.java to NavigationHeaderHolderFactory(),
        NavigationTextItem::class.java to NavigationTextItemHolderFactory(),
        NavigationSplitter::class.java to NavigationSplitterHolderFactory(),
        NavigationSubHeader::class.java to NavigationSubHeaderHolderFactory(),
        NavigationCheckBoxItem::class.java to NavigationCheckBoxItemHolderFactory()
    )

    init {
        inflater = context.getSystemService(Context.LAYOUT_INFLATER_SERVICE) as LayoutInflater
        this.items = flattenItems(items).toMutableList()
    }

    fun setNavigationItems(items: Array<NavigationItem>) {
        this.items.clear()
        this.items.addAll(flattenItems(items))
        notifyDataSetChanged()
    }

    override fun areAllItemsEnabled(): Boolean = false

    override fun isEnabled(position: Int): Boolean = items[position].enabled

    override fun getCount(): Int = items.size

    override fun getItem(position: Int): NavigationItem = items[position]

    override fun getItemId(position: Int): Long = position.toLong()

    override fun hasStableIds(): Boolean = false

    override fun getView(position: Int, convertView: View?, parent: ViewGroup): View {
        val holder: IHolder
        val view: View
        if (convertView == null) {
            val factory = viewItemHolderFactories[items[position].javaClass]!!
            view = factory.createView(inflater, parent)
            holder = factory.createHolder(view)
        } else {
            view = convertView
            holder = convertView.tag as IHolder
        }
        holder.update(items[position])
        return view
    }

    override fun getItemViewType(position: Int): Int = viewItemTypes[items[position].javaClass]!!

    override fun getViewTypeCount(): Int = viewItemTypes.size

    override fun isEmpty(): Boolean = items.isEmpty()

    private fun flattenItems(items: Array<NavigationItem>): List<NavigationItem> {
        val flattenList = mutableListOf<NavigationItem>()
        for (item in items) {
            flattenList.add(item)
            if (item is INavigationItemGroup) {
                flattenList.addAll(flattenItems(item.getItems()))
            }
        }
        return flattenList
    }
}
